File: mobile/src/constants/tokens.ts
// tokens.ts — Tally Design System v2 (2026 refresh): the single source of truth
// for the spatial, type, radius, colour and elevation scales. Ported from the
// handoff's tokens.css so engineering and design share one vocabulary.
//
// Notes on fidelity:
//   • Every spacing and metric in the app sits on the 4pt grid — padding,
//     margins, gaps, icon point sizes, corner radii, control dimensions. The
//     artboards are pixel-tuned and frequently weren't (7pt row padding, 9pt
//     radii, a 21.5pt ring); those were snapped to the nearest rung so the app
//     reads as one system rather than a set of traced screens.
//   • Two things are outside the grid by design: the type scale below, which is
//     its own progression (13.5pt note text is deliberate), and the Elevation
//     shadows, whose radii and offsets aren't layout quantities. Anywhere else,
//     an off-grid number needs a comment saying which system constant it's
//     cancelling — see MENU_PLATTER_PAD in components/tally/swipe-row.
//   • Type families are intentionally Geist / Geist Mono here, not the design's
//     System (SF) / Spline Sans Mono — see TallyFonts in tally-theme.ts for the
//     rationale. The numeric scale below is family-agnostic.

import type { TextStyle } from 'react-native';

/** Type scale — size in pt. Hierarchy leans on weight, not size (DS v2). */
export const TextScale = {
  displayXl: 34, // hero statement      — display
  displayLg: 28, // screen title        — display (≈ iOS Large Title)
  displayMd: 20, // tab name            — display
  numXl: 36, // live draft amount       — mono
  numLg: 22, // running total           — mono
  bodyLg: 16, // primary body / note
  bodyMd: 14, // default UI text
  bodySm: 13, // secondary / buttons
  caption: 12, // chips, fine print
  label: 11, // mono eyebrow / meta (uppercase, tracked)
  labelSm: 10, // mono micro-label
  micro: 9, // badges
} as const;

/** Letter-spacing tokens, expressed in pt for the size they pair with. */
export const Tracking = {
  tight: -0.02, // display, as a ratio of font size
  label: 0.14, // mono labels (em)
  eyebrow: 0.22, // mono eyebrows (em)
} as const;

/** Spacing — the 4px grid. Numeric (pt) so it drops straight into styles. */
export const Space = {
  s0: 0,
  s1: 4,
  s2: 8,
  s3: 12,
  s4: 16,
  s5: 20,
  s6: 24,
  s7: 28,
  s8: 32,
  s10: 40,
  s12: 48,
} as const;

/** Corner radii, on the same 4pt grid. `pill` is the fully-rounded sentinel. */
export const Radius = {
  xs: 4,
  sm: 8,
  md: 12,
  lg: 16,
  xl: 24, // glass cards / keypad keys
  pill: 999,
} as const;

/** Device canvas the artboards were laid out against (iPhone 15 logical pt). */
export const Frame = { width: 393, height: 852 } as const;

/** Cool-neutral neutrals + inverted "deep" surface, per theme. */
export const Neutral = {
  light: {
    screen: '#ececef',
    ink: '#1a1a1d',
    ink2: '#6b6b73',
    ink3: '#a6a6af',
    line: '#e4e4ea',
    card: '#ffffff',
    key: '#ffffff',
    keyLine: 'rgba(0,0,0,0.05)',
    deep: '#1b1b1e', // inverted surface (CTAs, total bar)
    deepInk: '#f4f4f6',
  },
  dark: {
    screen: '#161618',
    ink: '#f3f3f5',
    ink2: '#a4a4ac',
    ink3: '#65656e',
    line: '#2a2a2e',
    card: '#1f1f22',
    key: '#242428',
    keyLine: 'rgba(255,255,255,0.05)',
    deep: '#f3f3f5',
    deepInk: '#161618',
  },
} as const;

export type AccentToken = {
  name: string;
  accent: string;
  softLight: string;
  softDark: string;
  inkLight: string;
  inkDark: string;
};

/** Five interchangeable accents; Raspberry is the brand default. */
export const Accents: AccentToken[] = [
  { name: 'Raspberry', accent: '#b3476a', softLight: '#f4dbe2', softDark: '#3a1c26', inkLight: '#9c3458', inkDark: '#f0a8bd' },
  { name: 'Magenta', accent: '#c33a6e', softLight: '#f6d6e3', softDark: '#3a1722', inkLight: '#a82a5a', inkDark: '#f3a0bf' },
  { name: 'Plum', accent: '#864a7a', softLight: '#ecd9e8', softDark: '#2f2030', inkLight: '#6e3a64', inkDark: '#d6a6cd' },
  { name: 'Wine', accent: '#9c3a48', softLight: '#f1d6d9', softDark: '#341a1d', inkLight: '#852e3a', inkDark: '#e6a0a8' },
  { name: 'Ink', accent: '#2a2420', softLight: '#e6ddcf', softDark: '#2a2420', inkLight: '#2a2420', inkDark: '#d8cdbb' },
];

type Shadow = Pick<TextStyle, 'shadowColor' | 'shadowOpacity' | 'shadowRadius'> & {
  shadowOffset: { width: number; height: number };
  elevation: number;
};

/**
 * Elevation — the three lifts the design uses, as ready-to-spread RN shadow
 * props. `glass` is the frosted card/sheet lift, `cta` the tinted press under a
 * deep-ink button (pass the deep colour at the call site), `pop` the popover
 * float. CSS counterparts: --shadow-glass / --shadow-cta / --shadow-pop.
 */
export const Elevation = {
  glass: {
    shadowColor: '#000',
    shadowOpacity: 0.1,
    shadowRadius: 28,
    shadowOffset: { width: 0, height: 12 },
    elevation: 3,
  } as Shadow,
  cta: {
    shadowOpacity: 0.4,
    shadowRadius: 16,
    shadowOffset: { width: 0, height: 6 },
    elevation: 4,
  } as Omit<Shadow, 'shadowColor'>,
  pop: {
    shadowColor: '#0c0c14',
    shadowOpacity: 0.42,
    shadowRadius: 30,
    shadowOffset: { width: 0, height: 18 },
    elevation: 8,
  } as Shadow,
} as const;

File: app/Tally/Design/Tokens.swift
// Tokens.swift — Tally Design System v2: the spatial, type, radius, colour and
// elevation scales. Transcribed from mobile/src/constants/tokens.ts, keeping
// the names, so the handoff and the code still share one vocabulary.
//
// Notes on fidelity, carried over from the original:
//   • Every spacing and metric in the app sits on the 4pt grid — padding,
//     margins, gaps, icon point sizes, corner radii, control dimensions. Where
//     an off-grid number is unavoidable it needs a comment saying which system
//     constant it is cancelling.
//   • Two things are outside the grid by design: the type scale, which is its
//     own progression (13.5pt note text is deliberate), and the elevation
//     shadows, whose radii and offsets are not layout quantities.
//   • The families are Geist / Geist Mono rather than the design's System (SF)
//     and Spline Sans Mono — a deliberate divergence. See Fonts.swift.
import SwiftUI

/// Type scale — size in points. Hierarchy leans on weight, not size.
enum TextScale {
  static let displayXl: CGFloat = 34  // hero statement
  static let displayLg: CGFloat = 28  // screen title (≈ iOS Large Title)
  static let displayMd: CGFloat = 20  // tab name
  static let numXl: CGFloat = 36  // live draft amount — mono
  static let numLg: CGFloat = 22  // running total — mono
  static let bodyLg: CGFloat = 16  // primary body / note
  static let bodyMd: CGFloat = 14  // default UI text
  static let bodySm: CGFloat = 13  // secondary / buttons
  static let caption: CGFloat = 12  // chips, fine print
  static let label: CGFloat = 11  // mono eyebrow / meta (uppercase, tracked)
  static let labelSm: CGFloat = 10  // mono micro-label
  static let micro: CGFloat = 9  // badges
}

/// Letter spacing. `label` and `eyebrow` are ems — multiply by the font size.
enum Tracking {
  static let tight = -0.02  // display, as a ratio of font size
  static let label = 0.14  // mono labels (em)
  static let eyebrow = 0.22  // mono eyebrows (em)

  /// An em value at a given size, which is what SwiftUI's `.tracking` wants.
  static func pt(_ em: Double, at size: CGFloat) -> CGFloat { CGFloat(em) * size }
}

/// Spacing — the 4pt grid.
enum Space {
  static let s0: CGFloat = 0
  static let s1: CGFloat = 4
  static let s2: CGFloat = 8
  static let s3: CGFloat = 12
  static let s4: CGFloat = 16
  static let s5: CGFloat = 20
  static let s6: CGFloat = 24
  static let s7: CGFloat = 28
  static let s8: CGFloat = 32
  static let s10: CGFloat = 40
  static let s12: CGFloat = 48
}

/// Corner radii, on the same 4pt grid.
enum Radius {
  static let xs: CGFloat = 4
  static let sm: CGFloat = 8
  static let md: CGFloat = 12
  static let lg: CGFloat = 16
  static let xl: CGFloat = 24  // glass cards / keypad keys
  /// Fully rounded. In SwiftUI this is `Capsule()`, not a number — kept for
  /// the places that still take a radius.
  static let pill: CGFloat = 999
}

/// Cool-neutral neutrals plus the inverted "deep" surface, per theme.
enum Neutral {
  struct Palette {
    let screen: Color
    let ink: Color
    let ink2: Color
    let ink3: Color
    let line: Color
    let card: Color
    let field: Color
    let key: Color
    let keyLine: Color
    /// Inverted surface — CTAs, the total bar.
    let deep: Color
    let deepInk: Color
  }

  static let light = Palette(
    screen: Color(hex: "#f0f1f3"),
    ink: Color(hex: "#171719"),
    ink2: Color(hex: "#5f6066"),
    ink3: Color(hex: "#82838c"),  // ≥4:1 on the screen
    line: Color(hex: "#e3e4e8"),
    card: Color(hex: "#ffffff"),
    field: Color(hex: "#f6f6f8"),
    key: Color(hex: "#ffffff"),
    keyLine: Color(white: 0, opacity: 0.04),
    deep: Color(hex: "#171719"),
    deepInk: Color(hex: "#ffffff")
  )

  static let dark = Palette(
    screen: Color(hex: "#161618"),
    ink: Color(hex: "#f3f3f5"),
    ink2: Color(hex: "#a4a4ac"),
    ink3: Color(hex: "#8a8a93"),
    line: Color(hex: "#2a2a2e"),
    card: Color(hex: "#1f1f22"),
    field: Color(hex: "#26262a"),
    key: Color(hex: "#242428"),
    keyLine: Color(white: 1, opacity: 0.05),
    deep: Color(hex: "#f3f3f5"),
    deepInk: Color(hex: "#161618")
  )
}

/**
 Elevation — the three lifts the design uses.

 `glass` is the frosted card/sheet lift, `cta` the tinted press under a deep-ink
 button (the caller supplies the colour), `pop` the popover float. The React
 Native version spread these as shadow props; here they are a modifier, so the
 call sites read the same but the numbers live in one place.
 */
enum Elevation {
  case glass, cta(Color), pop

  var color: Color {
    switch self {
    case .glass: return Color.black.opacity(0.10)
    case .cta(let tint): return tint.opacity(0.40)
    case .pop: return Color(hex: "#0c0c14").opacity(0.42)
    }
  }

  var radius: CGFloat {
    switch self {
    case .glass: return 28
    case .cta: return 16
    case .pop: return 30
    }
  }

  var y: CGFloat {
    switch self {
    case .glass: return 12
    case .cta: return 6
    case .pop: return 18
    }
  }
}

extension View {
  func elevation(_ e: Elevation) -> some View {
    shadow(color: e.color, radius: e.radius, x: 0, y: e.y)
  }
}

// MARK: - Hex colours

extension Color {
  /**
   A `#rrggbb` (or `#rrggbbaa`) string, as the design tokens spell them.

   Kept because the palette crosses a boundary: `TallyKit.Accent` stores hex so
   the store can resolve a persisted — possibly retired — accent without
   depending on SwiftUI. This is where those strings become colours.
   */
  init(hex: String) {
    var s = hex
    if s.hasPrefix("#") { s.removeFirst() }
    var v: UInt64 = 0
    Scanner(string: s).scanHexInt64(&v)
    let r: Double
    let g: Double
    let b: Double
    var a: Double = 1
    switch s.count {
    case 8:
      r = Double((v >> 24) & 0xff) / 255
      g = Double((v >> 16) & 0xff) / 255
      b = Double((v >> 8) & 0xff) / 255
      a = Double(v & 0xff) / 255
    default:
      r = Double((v >> 16) & 0xff) / 255
      g = Double((v >> 8) & 0xff) / 255
      b = Double(v & 0xff) / 255
    }
    self.init(.sRGB, red: r, green: g, blue: b, opacity: a)
  }
}
